statsd/statsd.go

```diff
@@ -152,20 +138,51 @@ func (s *Server) Run(ctx context.Context) error {
 		percentThresholds = append(percentThresholds, pt)
 	}
 
-	aggregator := NewMetricAggregator(backends, percentThresholds, s.FlushInterval, s.ExpiryInterval, s.MaxWorkers, s.DefaultTags)
-	go aggregator.Aggregate()
-	s.aggregator = aggregator
+	aggregator := NewMetricAggregator(backends, percentThresholds, s.FlushInterval, s.ExpiryInterval, s.DefaultTags)
+	var wgAggr sync.WaitGroup
+	var wgReceiver sync.WaitGroup
+	wgAggr.Add(1)
+	go func() {
+		defer wgAggr.Done()
+		aggregator.Aggregate()
+	}()
+	defer func() {
+		wgReceiver.Wait()             // Wait for all receivers to finish
+		close(aggregator.MetricQueue) // Tell the aggregator to shutdown
+		wgAggr.Wait()                 // Wait for aggregator to shutdown
+	}()
 
 	// Start the metric receiver
-	f := func(metric *types.Metric) {
-		aggregator.MetricQueue <- metric
+	f := func(ctx context.Context, metric *types.Metric) {
+		select {
+		case aggregator.MetricQueue <- metric:
+		case <-ctx.Done():
+		}
 	}
 	cloud, err := cloudprovider.InitCloudProvider(s.CloudProvider, s.Viper)
 	if err != nil {
 		return err
 	}
-	receiver := NewMetricReceiver(s.MetricsAddr, s.Namespace, s.MaxReaders, s.MaxMessengers, s.DefaultTags, cloud, HandlerFunc(f))
-	go receiver.ListenAndReceive()
+
+	c, err := net.ListenPacket("udp", s.MetricsAddr)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		// This makes receivers error out and stop
+		if err := c.Close(); err != nil {
+			log.Warnf("Error closing socket: %v", err)
+		}
+	}()
+
+	receiver := NewMetricReceiver(s.Namespace, s.DefaultTags, cloud, HandlerFunc(f))
+	wgReceiver.Add(s.MaxReaders)
+	for r := 0; r < s.MaxReaders; r++ {
+		go func() {
+			defer wgReceiver.Done()
+			receiver.Receive(ctx, c)
+		}()
+	}
```

Comment:
We spawn `s.MaxReaders` pipelines here. They all run concurrently and read-process-enqueue metrics onto the aggregator's queue.